lib/name_gender_classifier.rb

```diff
@@ -0,0 +1,104 @@
+require 'iconv'
+
+# Gender detector for first names.
+module NameGenderClassifier
+  # Return the gender(s) (probabilistically) for the informed name(s). The result type will vary depending
+  # on the parameter type:
+  #
+  # [String, Symbol] the gender (String) is returned.
+  # [Array<String>] an array (Array<String>) with the genders is returned.
+  # [Array<Object>] an array (Array<Object>) with the same objects and the newly assigned genders is returned.
+  #
+  # @param arg [String, Symbol, Array<String>, Array<Symbol>, Array<Object>] argument holding first
+  #   name(s) information(s).
+  # @paran options [Hash] first_name_attribute: name of the method that returns the first name,
+  #                       gender_attribute: name of the method which will receive the gender assignment.
+  #
+  # @return [String, Array<String>, Array<Object>] the gender classification for the passed first names
+  def self.classify(arg, options = {})
+    case arg
+    when String, Symbol
+      most_probable_gender(arg)
+    when Array
+      if arg[0].is_a?(String) || arg[0].is_a?(Symbol)
+        classify_array(arg)
+      else
+        classify_objects(arg, options)
+      end
+    end
+  end
+
+  # Return the genders (probabilistically) for the informed names.
+  #
+  # @param array [Array<String>, Array<Symbol>] see {NameGenderClassifier.classify}
+  #
+  # @return [Array<String>] see {NameGenderClassifier.classify}
+  def self.classify_array(array)
+    result = []
+    DatabaseManager.gdbm do |db|
+      array.each do |name|
+        next unless name
+
+        result << most_probable_gender(name, db)
+      end
+    end
+
+    result
+  end
+
+  # For each object in the array, it tries to classify the gender for object.first_name or
+  # object.name (or equivalent method) and save it on object.gender (or equivalent method).
+  #
+  # @param objects [Array<Object>] see {NameGenderClassifier.classify}
+  # @param options see {NameGenderClassifier.classify}
+  #
+  # @return [Array<Object>] the objects with the assigned genders
+  def self.classify_objects(objects, options = {})
+    gender_attribute = options.fetch(:gender_attribute, 'gender')
+    first_name_attribute = options.fetch(:first_name_attribute, nil)
+    gender_attribute_assignment = "#{gender_attribute}="
+    if first_name_attribute.nil?
```

Comment:
I feel like there may be a way to implement the fallbacks to `first_name_attributes` without the if statement. 

What if you do: 

first_name_attribute = options.fetch(:first_name_attribute, nil) || (:first_name if defined?(objects[0].first_name)) || (:name if defined?(objects[0].name))